File: sem_7/computer_networks/partA_python/q2.py
# Node class contains 2 things -> distance and previous node
class Node:
    def __init__(self, total_nodes) -> None:
        self.dist = [float('inf')] * total_nodes
        self.previousNode = [None] * total_nodes

def bf(total_nodes, cost):
    node = [Node(total_nodes) for _ in range(total_nodes)]
    # fill up adj matrix
    for i in range(total_nodes):
        for j in range(total_nodes):
            node[i].dist[j], node[i].previousNode[j] = cost[i][j], j
    # perform bf
    for _ in range(total_nodes - 1):
        for i in range(total_nodes):
            for j in range(total_nodes):
                for k in range(total_nodes):
                    if node[i].dist[j] > node[i].dist[k] + node[k].dist[j]:
                        node[i].dist[j] = node[i].dist[k] + node[k].dist[j]
                        node[i].previousNode[j] = k
    # print
    for i in range(total_nodes):
        print(f"State value for router {i} is")
        for j in range(total_nodes):
            print(f"Via {node[i].previousNode[j]} Distance {node[i].dist[j]}")


def main():
    total_nodes = int(input("Enter number of nodes : "))
    cost = [list(map(int, input().split())) for _ in range(total_nodes)]
    bf(total_nodes, cost)
main()

File: sem_7/computer_networks/q2.c
#include <stdio.h>
#include <string.h>
struct node {
  int dist[20];
  int from[20];
} rt[10];

int main() {
  int dmat[20][20];
  int n, i, j, k, count = 0;
  printf("Enter The Number Of Nodes\n");
  scanf("%d", &n);
  printf("Enter The Cost Matrix\n");
  for (i = 0; i < n; i++)
    for (j = 0; j < n; j++) {
      scanf("%d", &dmat[i][j]);
      dmat[i][i] = 0;
      rt[i].dist[j] = dmat[i][j];
      rt[i].from[j] = j;
    }

  do {
    count = 0;
    for (i = 0; i < n; i++)
      for (j = 0; j < n; j++)
        for (k = 0; k < n; k++)
          if (rt[i].dist[j] > dmat[i][k] + rt[k].dist[j]) {
            rt[i].dist[j] = rt[i].dist[k] + rt[k].dist[j];
            rt[i].from[j] = k;
            count++;
          }
  } while (count != 0);
  for (i = 0; i < n; i++) {
    printf("\nState Value For Router %d Is\n", i);
    for (j = 0; j < n; j++) {
      printf("\t\tVia %d Distance %d\n", rt[i].from[j], rt[i].dist[j]);
    }
  }
}
